server/src/main/java/org/elasticsearch/action/support/nodes/TransportNodesAction.java

```diff
@@ -254,14 +254,28 @@ private void onFailure(int idx, String nodeId, Throwable t) {
         }
 
         private void finishHim() {
-            threadPool.executor(finalExecutor).execute(() -> newResponse(task, request, responses, listener));
+            if (isCancelled(task)) {
+                listener.onFailure(new TaskCancelledException("task cancelled"));
+                return;
+            }
+
+            final String executor = finalExecutor.equals(ThreadPool.Names.SAME) ? ThreadPool.Names.GENERIC : finalExecutor;
+            threadPool.executor(executor).execute(() -> newResponse(task, request, responses, listener));
         }
     }
 
-    class NodeTransportHandler implements TransportRequestHandler<NodeRequest> {
+    // Visible for testing, allows bypassing task cancellation handling
```

Comment:
This seems surprising - as far as I can tell there's only a couple of places in tests where this matters, but in both places we're actually testing something to do with cancelling a `TransportNodesAction`. I think those tests should be updated to exercise the new behaviour, rather than working around it like this.